### PR TITLE
Job tracking and new finalize Tasks

### PR DESCRIPTION
**Pull request by aarontp**
Some notes about this change:
- Sorry for the big change, I didn't anticipate it being this large, but a decent chunk of this is just new tests.
- Jobs are instantiated per request instead of keeping the same Job around for the lifecycle of the task manager.
- Tasks are now tracked by adding them to the instantiated Job that generated them instead of one big list.
- Evidence returned from tasks is saved in an EvidenceCollection object in the Job so that we can later pass it to the "finalize" tasks
- The task manager tracks all Job objects until the Request and the finalize tasks have fully completed.
- This adds the new concepts of Job finalization and Request finalization Tasks.  When a Job ends, `TurbiniaJob.create_final_task()` gets called, which can generate a final task to process all of the other output from the Job.  Likewise when all Jobs for the entire processing request have completed, a special RequestFinalizationJob is run and can create final new Tasks.  These can be used to combine multiple Task outputs into one output, and will allow Jobs to be broken up into more parallel Tasks.  These new Tasks that are generated get passed the entire EvidenceCollection for the given Job/Request.
- This adds a specific Request finalization Task that will generate and save the full report out to GCS, along with the associated metadata file.  This can be used to generate a GCS/pubsub trigger so Timesketch can import the report as a story.
- Renamed `finalize_result()` to `process_result()` to disambiguate between processing and finalizing.  Also refactored some of this out into `process_job()`. 
- Adds a new FinalReport Evidence type that is configured to save metadata file
- A new EvidenceCollection type is added, which extends Evidence so we can pass it around as evidence, and also has a `collection` list property so we can save other Evidence objects there.
- self.tasks becomes a @property instead of a manually managed list.
- Adds tests for the Task manager, and brings test coverage from 58% to 73%

### CODE REVIEW

**Comment by aarontp** [comment]
Ok, this one should be ready to review now, thanks!

File: turbinia/task_manager.py [hunk 1]
```diff
@@ -181,42 +192,164 @@ def add_evidence(self, evidence_):
...
+        '{1:s}'.format(request_id, final_job.id))
+
+    # Finalize tasks use EvidenceCollection with all evidence created by the
+    # request or job.
+    final_evidence = evidence.EvidenceCollection()
+    final_evidence.request_id = request_id
+    self.running_jobs.append(final_job)
+
+    # Gather evidence created by every Job in the request.
+    for running_job in self.running_jobs:
```
Comment (berggren) [inline turbinia/task_manager.py]: So the jobs are still in self.running_jobs even when all their tasks are done?

Comment (berggren) [reply turbinia/task_manager.py]: Ah, I see, so the running_jobs are still populated when the finalize task is added. Jobas are removed after the Finalize task har run. Correct?

File: turbinia/task_manager.py [hunk 2]
```diff
@@ -248,33 +387,83 @@ def finalize_result(self, task_result):
...
+    request_id = job.request_id
+    request_done = self.check_request_done(request_id)
+    request_finalized = self.check_request_finalized(request_id)
+    # If the request is done but not finalized, we generate the finalize tasks.
+    if request_done and not request_finalized:
+      self.generate_request_finalize_tasks(job)
+
+    # If the Job has been finalized then we can remove all the Jobs for this
+    # request since everything is complete.
+    elif request_done and request_finalized:
```
Comment (berggren) [inline turbinia/task_manager.py]: This answers my comment above :)

File: turbinia/workers/finalize_request.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015 Google Inc.
```
Comment (berggren) [inline turbinia/workers/finalize_request.py]: 2019

File: turbinia/evidence.py [hunk 4]
```diff
@@ -222,6 +227,33 @@ def validate(self):
...
+    super(EvidenceCollection, self).__init__(*args, **kwargs)
+    self.collection = collection if collection else []
+
+  def serialize(self):
+    """Return JSON serializable object."""
+    serialized_evidence = super(EvidenceCollection, self).serialize()
+    serialized_evidence['collection'] = [e.serialize() for e in self.collection]
+    return serialized_evidence
+
+  def add_evidence(self, evidence_):
```
[the remark this answers is not in the document]
Comment (aarontp) [reply turbinia/evidence.py]: Good question...  Fixed.

File: turbinia/jobs/finalize_request.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015 Google Inc.
```
[the remark this answers is not in the document]
Comment (aarontp) [reply turbinia/jobs/finalize_request.py]: copypasta fail.  Fixed, thanks.

File: turbinia/task_manager.py (hunk 1 above)
Comment (aarontp) [reply turbinia/task_manager.py]: Yeah, that's correct :)

File: turbinia/task_manager.py (hunk 2 above)
Comment (aarontp) [reply turbinia/task_manager.py]: :) Ok, cool.

File: turbinia/workers/finalize_request.py (hunk 3 above)
Comment (aarontp) [reply turbinia/workers/finalize_request.py]: Fixed.

**Comment by aarontp** [comment]
Thanks for the review!

**Pull request merged**